scripts/tybalt_vae.py

```diff
@@ -0,0 +1,320 @@
+
+
+# In[15]:
+
+# Visualize training performance
+history_df = pd.DataFrame(hist.history)
+hist_plot_file = os.path.join('figures', 'onehidden_vae_training.pdf')
+ax = history_df.plot()
+ax.set_xlabel('Epochs')
+ax.set_ylabel('VAE Loss')
+fig = ax.get_figure()
+fig.savefig(hist_plot_file)
+
+
+# ## Compile and output trained models
+# 
+# We are interested in:
+# 
+# 1. The model to encode/compress the input gene expression data
+#   * Can be possibly used to compress other tumors
+# 2. The model to decode/decompress the latent space back into gene expression space
+#   * This is our generative model
+# 3. The latent space compression of all pan cancer TCGA samples
+#   * Non-linear reduced dimension representation of tumors can be used as features for various tasks
+#     * Supervised learning tasks predicting specific gene inactivation events
+#     * Interpolating across this space to observe how gene expression changes between two cancer states
+# 4. The weights used to compress each latent node
+#   * Potentially indicate learned biology differentially activating tumors
+
+# ### Encoder model
+
+# In[16]:
+
+# Model to compress input
+encoder = Model(rnaseq_input, z_mean_encoded)
+
+
+# In[17]:
+
+# Encode rnaseq into the hidden/latent representation - and save output
+encoded_rnaseq_df = encoder.predict_on_batch(rnaseq_df)
+encoded_rnaseq_df = pd.DataFrame(encoded_rnaseq_df, index=rnaseq_df.index)
+
+encoded_rnaseq_df.columns.name = 'sample_id'
+encoded_rnaseq_df.columns = encoded_rnaseq_df.columns + 1
+encoded_file = os.path.join('data', 'encoded_rnaseq_onehidden_warmup_batchnorm.tsv')
+encoded_rnaseq_df.to_csv(encoded_file, sep='\t')
+
+
+# In[18]:
+
+# What are the most and least activated nodes
+top_active_nodes = encoded_rnaseq_df.sum(axis=0).sort_values(ascending=False)
+print(top_active_nodes.head(10))
+top_active_nodes.tail(10)
+
+
+# In[19]:
+
+# Example distribution of latent layer
+plt.figure(figsize=(6, 6))
+plt.scatter(encoded_rnaseq_df.iloc[:, 1], encoded_rnaseq_df.iloc[:, 2])
+
+
+# ### Decoder (generative) model
+
+# In[20]:
+
+# build a generator that can sample from the learned distribution
+decoder_input = Input(shape=(latent_dim, ))  # can generate from any sampled z vector
+_x_decoded_mean = decoder_to_reconstruct(decoder_input)
+decoder = Model(decoder_input, _x_decoded_mean)
+
+
+# In[21]:
+
+# How well does the model reconstruct the input RNAseq data
+input_rnaseq_reconstruct = decoder.predict(np.array(encoded_rnaseq_df))
+input_rnaseq_reconstruct = pd.DataFrame(input_rnaseq_reconstruct, index=rnaseq_df.index,
+                                        columns=rnaseq_df.columns)
+input_rnaseq_reconstruct.head(3)
```

Comment:
Thinking about this a bit more - I should dig deeper into this. It would be interesting to observe reconstruction by cancer-type. If the model reconstructs a cancer-type worse than others, I may need to rethink current batch training...